vault-tool/src/Network/VaultTool/KeyValueV2.hs

```diff
@@ -0,0 +1,237 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+{- | A library for working with Vault's KeyValue version 2 secrets engine
+
+ Unless otherwise specified, all IO functions in this module may
+ potentially throw 'HttpException' or 'VaultException'
+-}
+module Network.VaultTool.KeyValueV2 (
+    VaultSecretVersion (..),
+    VaultSecretVersionMetadata (..),
+    VaultSecretMetadata (..),
+    vaultWrite,
+    vaultRead,
+    vaultReadVersion,
+    vaultDelete,
+    vaultList,
+    isFolder,
+    vaultListRecursive,
+) where
+
+import Control.Applicative ((<|>))
+import Data.Aeson (
+    FromJSON,
+    ToJSON,
+    object,
+    parseJSON,
+    toJSON,
+    withObject,
+    (.:),
+    (.=),
+ )
+import Data.Text (Text)
+import qualified Data.Text as T
+import Data.Time (UTCTime)
+
+import Network.VaultTool.Internal (
+    newDeleteRequest,
+    newGetRequest,
+    newListRequest,
+    newPostRequest,
+    runVaultRequestAuthenticated,
+    runVaultRequestAuthenticated_,
+    withStatusCodes,
+ )
+import Network.VaultTool.Types (
+    VaultConnection,
+    Authenticated,
+    VaultMountedPath (..),
+    VaultSearchPath (..),
+    VaultSecretPath (..),
+ )
+
+data VaultSecretMetadata = VaultSecretMetadata
+    { _VaultSecretMetadata_leaseDuration :: Int
+    , _VaultSecretMetadata_leaseId :: Text
+    , _VauleSecretMetadata_renewable :: Bool
+    }
+    deriving (Show, Eq {- TODO Ord -})
+
+instance FromJSON VaultSecretMetadata where
+    parseJSON = withObject "VaultSecretMetadata" $ \v ->
+        VaultSecretMetadata
+            <$> v .: "lease_duration"
+            <*> v .: "lease_id"
+            <*> v .: "renewable"
+
+data VaultSecretVersion a = VaultSecretVersion
+    { vsvData :: a
+    , vsvMetadata :: VaultSecretVersionMetadata
+    }
+    deriving (Show)
+
+instance FromJSON a => FromJSON (VaultSecretVersion a) where
+    parseJSON = withObject "VaultSecretVersion" $ \v ->
+        VaultSecretVersion
+            <$> v .: "data"
+            <*> v .: "metadata"
+
+data VaultSecretVersionMetadata = VaultSecretVersionMetadata
+    { vsvmCreatedTime :: UTCTime
+    , vsvmDeletionTime :: Maybe UTCTime
+    , vsvmDestroyed :: Bool
+    , vsvmVersion :: Int
+    }
+    deriving (Show)
+
+instance FromJSON VaultSecretVersionMetadata where
+    parseJSON = withObject "VaultSecretVersionMetadata" $ \v ->
+        VaultSecretVersionMetadata
+            <$> v .: "created_time"
+            <*> (v .: "deletion_time" >>= parseOptionalDate)
```

Comment:
You can use the `.:?` operator

Reply:
I tried a few different approaches including `(.:?)` but failed to get that to work for me. The date is a string and rather than omitting it if it does not exist it is an empty string. Parsing the empty sting as `Nothing` was failing for me. If you have an idea of how to make that work, let me know.

Reply:
In that case, the `optional` combinator would work.  https://hackage.haskell.org/package/base-4.16.0.0/docs/Control-Applicative.html#v:optional

Reply:
I'm using `<|>` in the body of `parseOptionalDate`. Is there a cleaner way for me to write this?

Reply:
`optional` is perfect:
```
optional original = Just <$> original <|> pure Nothing
```

Reply:
2c42064fa561f294710ccd8b3ca290ace3d42d3f